File: General-DSA-Questions/Some Named Algo & Genral Ques/15_mthRootOfNumber.cpp
// Given two numbers N and M, find the Nth root of M. The nth root of a number M is defined as a number X when raised to the power N equals M.
// If the 'nth root is not an integer, return -1.

// Intiution:
// 1. Find a range/scale in which our answer lies.
// 2. For Nth root of M, the answer lies between 1 to M. So, our answer definitely lies in this range.
// 3. Now, we need to find maximum value of X such that X^N <= M.
// 4. Brute force approach would be to iterate from 1 to M and check for each number if its Nth power is equal to M. This would take O(M) time.
// 5. We can optimize this using binary search on the range 1 to M.

// Notes link : https://takeuforward.org/data-structure/nth-root-of-a-number-using-binary-search

// Approach:
// 1. Initialize two pointers, left = 1 and right = M.
// 2. While left <= right:
//    a. Calculate mid = left + (right - left) / 2.
//    b. Calculate mid^N using a helper function to avoid overflow.
//    c. If mid^N == M, return mid.
//    d. If mid^N < M, move the left pointer to mid + 1.
//    e. If mid^N > M, move the right pointer to mid - 1.
// 3. If no integer root is found, return -1.


#include <iostream>
using namespace std;

long long power(int mid, int n, int m) {
    long long result = 1;
    for (int i = 0; i < n; i++) {
        result *= mid;
        if (result > m) return result; // Early exit if result exceeds m
    }
    return result;
}

int nthRoot(int n, int m) {
    int left = 1;
    int right = m;

    while (left <= right) {
        int mid = left + (right - left) / 2;
        long long midPowerN = power(mid, n, m);

        if (midPowerN == m) {
            return mid; // Found the Nth root
        } else if (midPowerN < m) {
            left = mid + 1; // Move to the right half
        } else {
            right = mid - 1; // Move to the left half
        }
    }

    return -1; // Nth root not found
}

int main() {
    int n = 3; // Example: Find the 3rd root
    int m = 27; // Example: of 27
    int result = nthRoot(n, m);
    if (result != -1) {
        cout << "The " << n << "th root of " << m << " is: " << result << endl;
    } else {
        cout << "The " << n << "th root of " << m << " is not an integer." << endl;
    }
    return 0;
}

// For these type of problems we need to find a range/scale in which our answer lies and then perform binary search on that range to find the optimal answer.
// Before applying binary search, think of brute force which is linear search on the answer space, and then optimize it using binary search.

File: JS-DSA-Questions/Some Named Algo & Genral Ques/15_mthRootOfNumber.js
// Given two numbers N and M, find the Nth root of M. The Nth root of M is a number X such that
// X^N === M. If the Nth root is not an integer, return -1.

// Intuition:
// 1. The answer lies between 1 and M.
// 2. We want the largest X with X^N <= M (and equal to M for an exact integer root).
// 3. Brute force scanning 1..M is O(M); binary search on that range is O(log M).

// Notes link : https://takeuforward.org/data-structure/nth-root-of-a-number-using-binary-search

// Approach:
// 1. left = 1, right = M.
// 2. While left <= right: mid = (left + right) / 2, compute mid^N (with early exit if it exceeds M).
//    - If mid^N === M, return mid.
//    - If mid^N < M, left = mid + 1.
//    - If mid^N > M, right = mid - 1.
// 3. If nothing found, return -1.

// JavaScript note: C++ used `long long` in the power helper to avoid overflow. JS numbers are
// 64-bit doubles (safe up to 2^53). We still early-exit once the running product exceeds m, which
// both matches the original and keeps values small.
function power(mid, n, m) {
  let result = 1;
  for (let i = 0; i < n; i++) {
    result *= mid;
    if (result > m) return result; // early exit
  }
  return result;
}

function nthRoot(n, m) {
  let left = 1;
  let right = m;

  while (left <= right) {
    const mid = left + Math.floor((right - left) / 2);
    const midPowerN = power(mid, n, m);

    if (midPowerN === m) {
      return mid; // exact Nth root
    } else if (midPowerN < m) {
      left = mid + 1;
    } else {
      right = mid - 1;
    }
  }

  return -1; // no integer Nth root
}

const n = 3;
const m = 27;
const res = nthRoot(n, m);
console.log(res !== -1 ? `The ${n}th root of ${m} is: ${res}` : `The ${n}th root of ${m} is not an integer.`); // 3
